docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,217 @@
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `PrepareProposalRequest`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `PrepareProposalRequest`, the application will
+ignore _all_ transactions sent to it in the request and instead reap up to
+`PrepareProposalRequest.max_tx_bytes` from it's own mempool. There is no need to
+execute the transactions for validity as they have already passed CheckTx.
+
+### `ProcessProposal`
+
+The `ProcessProposal` ABCI method is relatively straightforward. It is responsible
+for ensuring validity of the proposed block containing transactions that were
+selected from the `PrepareProposal` step. In order to check validity of the proposed
+block, we must iterate over the list of transactions and execute them using the
+same mode/execution strategy as `CheckTx`. In other words, we execute each transaction
+using the AnteHandler only -- no messages are executed. However, we cannot just
+execute `CheckTx` again, because `BaseApp` already has a modified `checkState` at
+this point. So when executing `ProcessProposal`, we create a similar branched
+state, `processProposalState`, off of `deliverState`. Using `processProposalState`
+we execute the AnteHandler for each transaction. Note, the `processProposalState`
+is never committed and is completely discarded after `ProcessProposal` completes.
+
+We will only populate the `Status` field of the `ProcessProposalResponse` with
+`ACCEPT` if ALL the transactions were accepted as valid, otherwise we will
+populate with `REJECT`.
+
+### `DeliverTx`
+
+Since transactions are not truly removed from the app-side mempool during
+`PrepareProposal`, since `ProcessProposal` can fail and we do not want to lose
+transactions, we need to finally remove the transaction from the app-side mempool
+during `DeliverTx` since during this phase, the transactions are being included
+in the proposed block.
+
+Alternatively, we can keep the transactions as truly being removed during the
+reaping phase in `PrepareProposal` and add them back to the app-side mempool in
+case `ProcessProposal` fails.
+
+## Consequences
+
+### Backwards Compatibility
+
+ABCI++ is naturally not backwards compatible with prior versions of the Cosmos SDK
+and Tendermint. For example, an application that requests `PrepareProposalRequest`
+to the same application that does not speak ABCI++ will naturally fail.
+
+However, in the first phase of the integration, the existing ABCI methods as we
+know them today will still exist and function as they currently do.
```

Comment:
Maybe worth mentioning that an application that implemented `ProcessProposal` by accepting all blocks passed to it and implemented `PrepareProposal` by returning the list passed to it, modulo any excess bytes (see [max_tx_bytes](https://github.com/tendermint/tendermint/blob/master/spec/abci%2B%2B/abci%2B%2B_methods.md#parameters-and-types-1)) would work exactly the same way as it did pre-ABCI++.